lld/MachO/ObjC.cpp

```diff
@@ -320,3 +342,1007 @@ void objc::checkCategories() {
+  err += msg;
+  error(err);
+  return false; // Always return false for easy 'return registerError()' syntax.
+}
+
+// This is a template so that it can be used both for CStringSection and
+// ConcatOutputSection
+template <typename T>
+bool ObjcCategoryMerger::collectSectionWriteInfoFromIsec(
+    InputSection *isec, InfroWriteSection<T> &catWriteInfo) {
+  if (catWriteInfo.valid)
+    return true;
+
+  catWriteInfo.inputSection = &isec->section;
+  catWriteInfo.align = isec->align;
+  catWriteInfo.outputSection = dyn_cast_or_null<T>(isec->parent);
+
+  if (isec->relocs.size())
+    catWriteInfo.relocTemplate = isec->relocs[0];
+
+  if (!catWriteInfo.outputSection) {
+    std::string message =
+        "Unexpected output section type for" + isec->getName().str();
+    return registerError(message.c_str());
+  }
+
+  catWriteInfo.valid = true;
+
+  return true;
+}
+
+bool ObjcCategoryMerger::tryGetSymbolAtIsecOffset(ConcatInputSection *isec,
+                                                  uint32_t offset,
+                                                  Symbol *&sym) {
+  const Reloc *reloc = isec->getRelocAt(offset);
+
+  if (!reloc)
+    return false;
+
+  sym = reloc->referent.get<Symbol *>();
+  return sym != nullptr;
+}
+
+bool ObjcCategoryMerger::tryGetDefinedAtIsecOffset(ConcatInputSection *isec,
+                                                   uint32_t offset,
+                                                   Defined *&defined) {
+  Symbol *sym;
+  if (!tryGetSymbolAtIsecOffset(isec, offset, sym))
+    return false;
+
+  defined = dyn_cast_or_null<Defined>(sym);
+  return defined != nullptr;
+}
+
+// Given an ConcatInputSection and an offset, if there is a symbol(Defined) at
+// that offset, then erase the symbol (mark it not live) from the final output.
+// Used for easely erasing already merged strings, method lists, etc ...
+bool ObjcCategoryMerger::tryEraseDefinedAtIsecOffset(ConcatInputSection *isec,
+                                                     uint32_t offset,
+                                                     bool stringOnly) {
+  const Reloc *reloc = isec->getRelocAt(offset);
+
+  if (!reloc)
+    return false;
+
+  Defined *sym = dyn_cast_or_null<Defined>(reloc->referent.get<Symbol *>());
+
+  if (!sym)
+    return false;
+
+  auto *cisec = dyn_cast_or_null<ConcatInputSection>(sym->isec);
+  if (!stringOnly && cisec) {
+    cisec->linkerOptimizeReason = LinkerOptReason::CategoryMerging;
+    return true;
+  }
+
+  if (auto *cisec = dyn_cast_or_null<CStringInputSection>(sym->isec)) {
+    uint32_t totalOffset = sym->value + reloc->addend;
+    StringPiece &piece = cisec->getStringPiece(totalOffset);
+    piece.linkerOptimizeReason = LinkerOptReason::CategoryMerging;
+    return true;
+  }
+
+  return false;
+}
+
+bool ObjcCategoryMerger::collectCategoryWriterInfoFromCategory(
+    InfoInputCategory &catInfo) {
+
+  if (!collectSectionWriteInfoFromIsec<ConcatOutputSection>(
```

Comment:
`error` is not fatal - it just outputs a message and then lld at the end returns an error code. 
For example you can have:
```
  error("error1");
  error("error2");
  ```
  and get:
```
ld64.lld: error: error1
ld64.lld: error: error2
```
The code is structured so that we only print out one error and after that return all the way out of category checker. This is because it's likely that the next error messages would be invalid / confusing - only the first error matters after which the whole category checker is put in a bad state. 

so we do have to have something like 
```
if ( false )
  return error("")
```
or
```
if ( false )  {
  error("")
  return false
}
```

I think options are:
1. don't output just one error - output and ignore errors (will probably output a few errors and crash later on)
2. hard system.exit() after first error - I don't think this is recommended in lld
3. leave as-is with first format above
4. refactor to 2nd format above

Or ... any other suggestions ?

Reply:
Yeah, it appears we shouldn't run `fatal()` per https://github.com/llvm/llvm-project/blob/main/lld/include/lld/Common/ErrorHandler.h#L32
But I still think there are many ways to simplify the logic with default value without checking boolean for each case. Eventually, `errorCount()` can be used in `Driver.cpp`.

Reply:
> But I still think there are many ways to simplify the logic with default value without checking boolean for each case.

I think this this is option 1 above where we just try to run as much as possible - though it will probably crash and not reach `Driver.cpp`.  
> If the check is too obvious (not expected from user input)

Yes, I do use asserts in several places where the failure would assert-worthy. Should I just make everything asserts ? Issue with that is that it may end up with no errors / possibly corrupt data or ambiguous crashes in release builds.